example/common/flat_write_stream.hpp

```diff
@@ -0,0 +1,554 @@
+
+                            op(boost::beast::error_code{}, 0);
+                            return init.result.get();
+                        } else {
+                            try {
+                                auto buf = std::shared_ptr<char>(new char[size], std::default_delete<char[]>());
+                                auto copy_size = boost::asio::buffer_copy(boost::asio::buffer(buf.get(), size),
+                                                                          buffers);
+                                write_some_op<
+                                        ConstBufferSequence,
+                                        boost::beast::handler_type<
+                                                WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1(buf.get(), copy_size), buf);
+
+                                op(boost::beast::error_code{}, 0);
+                                return init.result.get();
+                            }
+                            catch (std::bad_alloc const &) {
+                                // todo
+                                return;
+                            }
+                        }
+                    }
+                }
+                else {
+                    return next_layer_.async_write_some(buffers, handler);
+                }
+            }
+
+            friend
+            void
+            teardown(websocket::role_type,
+                     flat_write_stream&, boost::system::error_code& ec)
+            {
+                ec.assign(0, ec.category());
+            }
+
+            template<class TeardownHandler>
+            friend
+            void
+            async_teardown(websocket::role_type,
+                           flat_write_stream& s, TeardownHandler&& handler)
+            {
+                std::cout << "async_teardown" << std::endl;   //todo: remove
+
+                s.get_io_service().post(
+                        bind_handler(std::move(handler),
+                                     error_code{}));
+            }
+        };
+
+
+
+        template<class Stream>
+        template<class ConstBufferSequence, class Handler>
+        class flat_write_stream<
+                Stream>::write_some_op
+        {
+            Handler h_;
+            flat_write_stream& s_;
+            boost::asio::const_buffers_1 cb1_;
+            std::shared_ptr<char> ptr_;
+
+            int step = 0;
+
+        public:
+            write_some_op(write_some_op&&) = default;
+            write_some_op(write_some_op const&) = default;
+
+            template<class DeducedHandler, class... Args>
+            write_some_op(DeducedHandler&& h,
+                          flat_write_stream& s,
+                          boost::asio::const_buffers_1 cb1,
+                          std::shared_ptr<char> ptr)
```

Comment:
The signature could be:
```
template<
    class DeducedHandler,
    class ConstBufferSequence>
write_some_op(
    DeducedHandler&& h,
    flat_write_stream& s,
    ConstBufferSequence const& buffers);
```

And then just create your shared pointer and do the buffer copy inside the constructor.